[refactor] #4339: refactor `iroha_crypto`, update API

0x009922 opened the pull request:
## Description

See linked issue.

### Linked issue

Closes #4339

DCNick3 commented on crypto/src/signature/mod.rs:
```diff
@@ -633,4 +658,27 @@ mod tests {
+    #[test]
+    fn signature_serialized_representation() {
+        let input = json!({
+            "public_key": "e701210312273E8810581E58948D3FB8F9E8AD53AAA21492EBB8703915BBB565A21B7FCC",
```
The fact that pubkey is serialized as uppercase (mostly?), but payload is lowercase is not very nice..

Can we make it consistent?

0x009922 replied:
This goes back to https://github.com/hyperledger/iroha/pull/3316. TBH, I don't understand how this

> improves private key validation

@mversic, could you comment on that?

mversic replied:
> This goes back to https://github.com/hyperledger/iroha/pull/3316. TBH, I don't understand how this
> 
> > improves private key validation
> 
> @mversic, could you comment on that?

This was referring to something else. Uppercase/lowercase has no influence on validation

I would prefer making it all uppercase

0x009922 replied:
Changed `PrivateKey`'s and `Signature`'s `payload` serialized hex to uppercase.

I haven't bothered to updated every single hex in the workspace to uppercase accordingly. Hope it's fine.

mversic commented on crypto/src/signature/mod.rs:
```diff
@@ -82,10 +81,32 @@ impl Signature {
+    ///
+    /// This method exists to allow reproducing the signature in a more efficient way than through
+    /// deserialization.
+    pub fn from_raw(public_key: PublicKey, payload: &[u8]) -> Self {
```
please rename to `from_bytes` elsewhere also

0x009922 replied:
Done

mversic commented on crypto/src/lib.rs:
```diff
@@ -697,16 +728,9 @@ impl<'de> Deserialize<'de> for PrivateKey {
-            .map_err(D::Error::custom)
+        let PrivateKeySerialized { algorithm, payload } =
+            PrivateKeySerialized::deserialize(deserializer)?;
+        Self::from_raw(algorithm, &payload).map_err(D::Error::custom)
```
check out `serde(try_from` in `data_model/src/account.rs`. Is that applicable in this case?

0x009922 replied:
Thanks for pointing this out. `try_from` and `into` work perfectly here.

0x009922 commented:
Okay, so what's next:

- rename to `from_bytes` anyway
- change serialization to uppercase everywhere
- create an issue with a bug of `KeyGenConfiguration::from_private_key()`

0x009922 commented:
I've updated `KeyGenConfiguration` using type-state-pattern, but it fails in checks and with ffi features enabled.

Will address it next. Probably, will try to simplify the builder yet remain it type-safe.

mversic commented on crypto/src/lib.rs:
```diff
@@ -119,31 +120,54 @@ pub enum KeyGenOption<K = PrivateKey> {
+
+    impl From<Default> for Algorithm {
+        fn from(_: Default) -> Self {
+            use std::default::Default;
```
```suggestion
            #[cfg(not(feature = "std"))]
            use std::default::Default;
```

mversic commented on crypto/src/lib.rs:
```diff
@@ -119,31 +120,54 @@ pub enum KeyGenOption<K = PrivateKey> {
+    }
+
+    #[derive(Copy, Clone, Debug)]
+    pub struct Set(pub Algorithm);
```
```suggestion
    pub struct Custom(pub Algorithm);
```

maybe?

mversic replied:
I believe I can help you work around FFI macros.

0x009922 replied:
There was a problem that `KeyPair::generate_with_configuration` started to have a generic parameter to accept `KeyGenConfig<A>`, and FFI macros said generics aren't supported.

And this piece errored with a missing generic argument. Didn't succeed with adding one.

https://github.com/hyperledger/iroha/blob/e79f617f5edc53621e4561f057df92b425beeaca/crypto/src/lib.rs#L867-L873

mversic replied:
| see, we can't have generic objects in FFI.

in this case, you should have specified all concrete types like this:
```rs
 iroha_ffi::handles! { 
     KeyPairGenConfig<Default>, 
     KeyPairGenConfig<Set>,
     PublicKey, 
     PrivateKey, 
     KeyPair, 
     Signature, 
 }
 ```

mversic replied:
but then I'm not sure what would have happened with generic functions. I think there was only one potentially problematic `generate_with_configuration`. If you want to go back to this, tell me and we can work something out

0x009922 replied:
It's fine, I removed `with_algorithm` method and this problem is gone.

mversic replied:
I think you found a very good alternative

0x009922 commented on crypto/src/lib.rs:
```diff
@@ -119,31 +120,54 @@ pub enum KeyGenOption<K = PrivateKey> {
+
+    impl From<Default> for Algorithm {
+        fn from(_: Default) -> Self {
+            use std::default::Default;
```
Not relevant anymore.

0x009922 commented:
Updated how keys generation is configured. With some refactoring, managed to avoid unnecessary validation and panics in runtime thanks to types.

Renamed `KeyGenConfiguration` and `KeyGenOption` to `KeyPairGenConfig` and `KeyPairGenOption` respectively.

mversic commented on crypto/src/lib.rs:
```diff
@@ -110,66 +111,73 @@ impl FromStr for Algorithm {
-        /// Algorithm
-        algorithm: Algorithm,
+    #[cfg_attr(not(feature = "ffi_import"), derive(Debug))]
+    pub struct KeyPairGenConfig{
```
```suggestion
    pub struct KeyGenConfig{
```

I think previous name was fine

mversic commented on crypto/src/lib.rs:
```diff
@@ -110,66 +111,73 @@ impl FromStr for Algorithm {
+/// to pass it to a specific algorithm.
+/// Parameter `K` is intended to be set as a `PrivateKey` type of that specific algorithm.
+#[derive(Debug)]
+pub enum KeyPairGenOption<K> {
```
```suggestion
pub enum KeyGenOpt<K> {
```

I think the previous name was fine?

mversic replied:
there is no ambiguity

0x009922 replied:
I think so. Let me revert.

mversic commented on crypto/src/lib.rs:
```diff
@@ -224,21 +230,42 @@ impl KeyPair {
-            Algorithm::BlsNormal => signature::bls::BlsNormal::keypair(key_gen_option).into(),
-            Algorithm::BlsSmall => signature::bls::BlsSmall::keypair(key_gen_option).into(),
+    /// Generates a pair of Public and Private key with the corresponding [`KeyGenOption`].
+    pub fn generate_with_config(config: KeyGenConfig) -> Self {
```
instead of having `KeyGenConfig` I think it would be preferable to have 3 methods on `KeyPair`:
```rs
pub fn rand() -> Self;
pub fn from_seed(algorithm: Algorithm, seed: Vec<u8>) -> Self;
impl TryFrom<PrivateKey> for KeyPair {
    //..
}
```

or similar naming

what do you think?

0x009922 replied:
Yup, I think it would work totally fine and simplify things:

- `KeyPair::generate_from_random(algorithm)`
- `KeyPair::generate_from_seed(seed, algorithm)`
- `KeyPair::generate_from_private_key(private_key)`

Let me impl it tomorrow.

0x009922 replied:
And yeah, `TryFrom<PrivateKey>` makes sense too, although I still having a method for it is not bad.

Also, I think it would make more sense to use **"derive"** term:

- generate from random (or just "generate")
- derive from seed
- derive from private key

mversic replied:
> And yeah, TryFrom<PrivateKey> makes sense too, although still having a method for it is not bad.

I wouldn't be so eager to increase API surface. This can always be done so, as a compromise, I would make this into a separate issue to discuss and atm just leave `TryFrom`

> Also, I think it would make more sense to use "derive" term:

"derive" doesn't sound quite right to me. I'd vote for either:
1.
```rs
Keypair::random(algorithm: Algorithm)
Keypair::from_seed(algorithm: Algorithm, seed: Vec<u8>) -> Self;
```
2.
```rs
KeyPair::generate_from_random(algorithm: Algorithm)
KeyPair::generate_from_seed(algorithm: Algorithm, seed: Vec<u8>)
```

but I don't find `generate` prefix brings anything but a longer method name

0x009922 replied:
AFAIK, _derivation_ is an established term in cryptography. It applies to both deriving a key pair using pRNG + seed, and to deriving of the public key from a private one.

Some references:

- https://en.wikipedia.org/wiki/Key_derivation_function
- https://developer.mozilla.org/en-US/docs/Web/API/SubtleCrypto/deriveKey

Anyway, `KeyPair::random` and `KeyPair::from_seed` works fine, imo. And will leave `TryFrom` as the only way for `PrivateKey`.

0x009922 commented on crypto/src/lib.rs:
```diff
@@ -110,66 +111,73 @@ impl FromStr for Algorithm {
-        /// Algorithm
-        algorithm: Algorithm,
+    #[cfg_attr(not(feature = "ffi_import"), derive(Debug))]
+    pub struct KeyGenConfig{
+        inner: KeyGenConfigInner
```
Because `KeyGenConfigInner` is a enum, and it is also restricted in FFI to have either unit or single-tuple variants. So I introduced a layer of indirection here to satisfy FFI macros with `KeyGenConfig`.

0x009922 commented on crypto/src/lib.rs:
```diff
@@ -186,28 +136,64 @@ ffi::ffi_item! {
-    /// If public and private key don't match, i.e. if they don't make a pair
-    pub fn new(public_key: PublicKey, private_key: PrivateKey) -> Result<Self, Error> {
+    /// If public and private keys don't match, i.e. if they don't make a pair
+    pub fn from_raw_parts(public_key: PublicKey, private_key: PrivateKey) -> Result<Self, Error> {
```
See https://github.com/hyperledger/iroha/issues/4339#issuecomment-1984711378

I finally renamed it to `from_raw_parts`, because there is a mirrored method called `into_raw_parts` and it makes sense.

mversic replied:
the reason why we need a destructuring method is because we want to keep fields of our structs private but also not be required to clone if we want to move a field out of a struct. This is the same in `iroha_crypto` and `iroha_data_model` so we better find a general solution to this. And having constructors named `from_raw_parts` is not what I would like


IMO `raw_parts` implies some sort of raw pointers when you look at `std`. Instead of `KeyPair::into_raw_parts` we already have:
```rs
impl From<KeyPair> for (PublicKey, PrivateKey)
```

so we can just remove `KeyPair::into_raw_parts`. We can also just remove `KeyPair::new` and add:
```rs
impl From<(PublicKey, PrivateKey)> for KeyPair
```

mversic replied:
let's have a discussion on this at standup

mversic replied:
- let's keep `KeyPair::new`
- rename `KeyPair::into_raw_parts` into `KeyPair::into_parts`
- remove `From<KeyPair> for (Pk, Pk)`

0x009922 replied:
Addressed in [`ec0fa3a` (#4341)](https://github.com/hyperledger/iroha/pull/4341/commits/ec0fa3a1a4185f11deabb2cffa5c895a50f7569e), and created https://github.com/hyperledger/iroha/issues/4360 as discussed
